### DISCUSSION TITLE
Equality checks for NSIDs

### DISCUSSION THREAD

**Discussion by stevenbenitez** [bluesky-social/atproto#1141] [How to do it?] [answered]
The [spec for NSIDs](https://atproto.com/specs/nsid) permits upper and lowercase letters for both the authority and name segments. Given that domain names are case-insensitive and generally normalized to lowercase, should the following two NSIDs be considered equal?

```
io.social.getFeed
IO.SOCIAL.getFeed
```

Are both of these legal NSIDs?

I assume that the name should be treated as case-sensitive, since it may identify a method and most languages will permit both `getFeed` and `getfeed`.

**Comment by bnewbold** [comment, marked as answer]
Good question! We are working on formalizing our NSID specification a bit.

Current thinking is:

* the "domain" parts should be normalized to lower-case (including when the NSID is embedded in an AT-URI)
* the "name" part is case-sensitive (should not be normalized)
* for equality and uniqueness checks, case can be ignored on the "name" part, and thus the NSID as a whole
* there "must" never be multiple schemas from the same authority that differ only by casing (though unclear how this is enforced technically)
* there "should" never be multiple collections in the same atproto repository (aka, account storage) that differ only by case (again, the burden of enforcement/implementation is unclear)

The NSIDs embedded in record paths, inside records, and inside AT-URIs inside records, are the most important parts to get correct.

**Comment by stevenbenitez** [comment]
Thanks for the detailed reply, @bnewbold! What you outlined sounds reasonable. I'll proceed with treating the example NSIDs in my original post as being equal (case-insensitive comparison).

**Comment by bnewbold** [comment]
We addressed these questions, at least indirectly, in an updated spec document: https://atproto.com/specs/nsid

Going to convert this issue to a discussion, as it is helpful to keep around!
